Sim/SimDelphesInterface/src/DelphesSimulation.cpp

```diff
@@ -341,80 +409,516 @@ StatusCode DelphesSimulation::finalize() {
+
+    // Get corresponding jet constituents
+    for (auto itCand=cand->GetCandidates()->begin(); itCand!=cand->GetCandidates()->end(); ++itCand) {
+
+      // Jet constituent
+      Candidate* jetPart = static_cast<Candidate*>(*itCand);
+
+      // Get related MC particle recursively (different level of particle -> particle -> ... -> MC particle relations)
+      // Add index to the reference index field to avoid double counting
+      // Recursive procedure stops after the relation is to MC particle and not to a particle object or if particle not related to MC particle (<0 value)
+      findJetPartMC(jetPart, colMCParticles->size(), idRefMCPart);
+    } // Jet constituents
+
+    // Debug: print variable
+    double totSimE = 0;
+
+    for (auto id : idRefMCPart) {
+
+      auto relation = ascColJetsToMC->create();
+      relation.Jet(jet);
+      relation.Particle(colMCParticles->at(id));
+
+      // Debug: print FCC-EDM jet relation info
+      if (msgLevel() <= MSG::DEBUG) {
+        double recE   = sqrt(relation.Jet().Core().P4.Px*relation.Jet().Core().P4.Px +
+                             relation.Jet().Core().P4.Py*relation.Jet().Core().P4.Py +
+                             relation.Jet().Core().P4.Pz*relation.Jet().Core().P4.Pz +
+                             relation.Jet().Core().P4.Mass*relation.Jet().Core().P4.Mass);
+        double simE   = sqrt(relation.Particle().Core().P4.Px*relation.Particle().Core().P4.Px +
+                             relation.Particle().Core().P4.Py*relation.Particle().Core().P4.Py +
+                             relation.Particle().Core().P4.Pz*relation.Particle().Core().P4.Pz +
+                             relation.Particle().Core().P4.Mass*relation.Particle().Core().P4.Mass);
+        totSimE += simE;
+        debug() << " RefId: " << std::setw(3)            << id+1
+                << " Rel E: " << std::setprecision(2)
+                              << std::scientific
+                              << std::setw(9) << simE    << " "
+                              << std::setw(9) << totSimE << " <-> "
+                              << std::setw(9) << recE
+                              << std::fixed
+                              << std::endl;
+      } // Debug
+    }
+
+    // Debug: print end-line
+    if (msgLevel() <= MSG::DEBUG) debug() << endmsg;
+  } // For - jets
+}   
+
+//
+// Recursive method to find id of MCParticle related to the given jet Delphes Candidate object
+//
+void DelphesSimulation::findJetPartMC(Candidate* jetPart, int rangeMCPart, std::set<int>& idRefMCPart) {
+
+  // Recursion depth - increase
+  //static int depth = 0;
+  //depth++;
+
+  // Warning - no MC relation found
+  if (jetPart->GetCandidates()->GetEntries()==0) {
+
+    warning() << "Can't build one of the relations from Jet to MC particle!" << std::endl;
+  }
+  // Relation can be found
+  else for (auto itCand=jetPart->GetCandidates()->begin(); itCand!=jetPart->GetCandidates()->end(); ++itCand) {
```

Comment:
GetCandidates returns a pointer, so there is no CPU overhead for direct call. Curl braces will be added.